### ISSUE TITLE
Sourcegraph not recognizing that GitHub Enterprise repos are archived

### ISSUE THREAD

**Issue by dadlerj** [sourcegraph/sourcegraph#12549]
- **Sourcegraph version:** 3.16

Reported by https://app.hubspot.com/contacts/2762526/company/693777177

>When searching repositories on our internal Sourcegraph client, the archived: flag doesn’t seem to be reflected in search results

The customer reports seeing the same set of repos with `archived:yes` and `archived:no` (and no results at all with `archived:only`) even though at least one is archived (screenshot from GitHub Enterprise, the code host, confirms it). 

Do we recognize archived status from all code hosts, including GitHub Enterprise? Should I ask about the external service config here?

**Comment by keegancsmith** [comment]
We do, for example https://bigdata.sgdev.org is syncing for GHE and has plenty of archived repos: https://bigdata.sgdev.org/search?q=archived:true&patternType=literal

To debug I think we need to know the GHE version and see the external service config. But we do explicitly ask for the `isArchived` field from GHE's graphql endpoint. And when doing rest requests we look at the `archived` field of a repo.

Out of interest if they could return the output of this sql query for an archived repo that would be good:

```sql
select archived, metadata from repo where name like '%my/archived/repo%';
```

**Comment by dadlerj** [comment]
Confirmed they are using a GitHub config, and they are on the GHE 2.21.2 release from a few weeks ago.

Interestingly, their ext service config actually had the following in it:

```
"exclude": [
                {
                    "archived": true
                },
```

Asked them to run the query above.

**Comment by dadlerj** [comment]
And the result of that query:

```
sourcegraph=# select archived, metadata from repo where name like '%myorg/myrepo%';
 archived |                                                                                                                                  metadata
----------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 f        | {"ID": "$REDACTED", "URL": "https://github.example.com/myorg/myrepo", "IsFork": false, "IsPrivate": false, "DatabaseID": $REDACTED, "IsArchived": false, "Description": "", "NameWithOwner": "myorg/myrepo", "ViewerPermission": ""}
(1 row)
```

(actual org and repo name and URL removed)

**Comment by dadlerj** [comment]
But from the GHE instance: 
![image](https://user-images.githubusercontent.com/5589410/88946270-e1228880-d243-11ea-9070-0f8356b25559.png)

**Comment by dadlerj** [comment]
From the customer:
>One other thing I thought I’d mention: The fork exclusion on our instance does work, fork:yes/no/only all produce results as expected.

Perhaps there's a new version of the API for archived status? Or perhaps if the users explicitly `exclude` those repos in their ext service config, we automatically mark all repos as `archived=false`?

**Comment by tsenart** [comment]
@dadlerj: If they have `exclude: [{"archived": true}]`, then we should not sync any archived repos. It'd be good to see what the updated_at column for that repo shows — it might be that the syncer is taking a while and hasn't removed that repo yet.

**Comment by dadlerj** [comment]
Is it possible that there's some race-condition-y thing where the repo got put into a weird state? E.g.:

* We had the repo (before it was archived) on Sourcegraph
* They archived it
* Repo-updater checks for the repo list... The external service config tells it that we should exclude archived repos, and the GitHub API tells us that that repo was archived, so we stop fetching it and stop updating its metadata (WITHOUT going and flipping the "archived" flag in the db)
* The syncer doesn't realize that it is archived, so it keeps getting re-scheduled for syncing (without ever updating the repo metadata)

This is a totally made up image of how these things all work together, but I hope you can get the idea that I'm going for here :)

**Comment by unknwon** [comment]
> Is it possible that there's some race-condition-y thing where the repo got put into a weird state? E.g.:
> 
> * We had the repo (before it was archived) on Sourcegraph
> * They archived it
> * Repo-updater checks for the repo list... The external service config tells it that we should exclude archived repos, and the GitHub API tells us that that repo was archived, so we stop fetching it and stop updating its metadata (WITHOUT going and flipping the "archived" flag in the db)
> * The syncer doesn't realize that it is archived, so it keeps getting re-scheduled for syncing (without ever updating the repo metadata)
> 
> This is a totally made up image of how these things all work together, but I hope you can get the idea that I'm going for here :)

@dadlerj: This scenario makes total sense to me, so knowing what's the value of `updated_at` column for that specific repo would help (as @tsenart suggested).

```sql
select archived, metadata, updated_at from repo where name like '%my/archived/repo%';
```

**Comment by keegancsmith** [comment]
Yes `updated_at` would be useful here. I suspect the repository syncing may be stalled? Are new repos appearing in Sourcegraph? If they explicitly exclude a repo (as a test), does it get excluded? That would rule out syncing not running vs us not respecting the archived field.

Note: If a repository isn't returned by a source, it will be deleted. So if it's excluded it will get removed. This points at either syncing not running successfully, or us not parsing the correct field from the GHE API. 

> Perhaps there's a new version of the API for archived status?

I think this might where the conflict is coming from. It would be surprising, only so many ways to spell "archived" in an API response :)

> And the full API output (REST API) confirms that it is archived: true on GitHub.

Can you share that output?

**Comment by dadlerj** [comment]
@keegancsmith @unknwon find much more context in the customer thread in Slack: https://sourcegraph.slack.com/archives/CFNURMWHE/p1596125200035400?thread_ts=1596072731.028300&cid=CFNURMWHE

**Comment by keegancsmith** [comment]
I took a quick look at this, I'm not sure why it isn't working. Most likely reason is to me is still syncing isn't running for some reason, but it could be something else. @unknwon do you have bandwidth to dive deeper?

In the thread the updated_at time is `2020-06-08 18:01:39.318373+00` which indicates this repo hasn't changed in a while. But we don't update unless we think it has changed. I investigated the following and came up short:

1. We correctly unmarshal rest archived. I tested with a mini-program using the structs we unmarshal with against the user supplied example response. 
2. We detect archived changing in the syncer. https://github.com/sourcegraph/sourcegraph/blob/0a19e43b9ea405f2ba4aa54ce7aaf6a39a121bf4/cmd/repo-updater/repos/types.go#L627
3. We handle fork the same way we handle archived in store.go

I wish we had a simple way to tell the last time a full sync ran, I don't think we log that. This is a somewhat hacky way to find when we last updated a repo when it wasn't an insert. (we special case inserting so syncing could still be broken even though new repos are added):

```sql
select max(updated_at) from repo where updated_at = created_at;
```

I am guessing for the archived repo including all sql columns may help debugging.

Mini-program to test unmarshalling:

<details>

```go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
)

type restRepositoryPermissions struct {
	Admin bool `json:"admin"`
	Push  bool `json:"push"`
	Pull  bool `json:"pull"`
}

type restRepository struct {
	ID          string `json:"node_id"` // GraphQL ID
	DatabaseID  int64  `json:"id"`
	FullName    string `json:"full_name"` // same as nameWithOwner
	Description string
	HTMLURL     string `json:"html_url"` // web URL
	Private     bool
	Fork        bool
	Archived    bool
	Permissions restRepositoryPermissions `json:"permissions"`
}

func main() {
	b, err := ioutil.ReadFile("/tmp/foo.json")
	if err != nil {
		log.Fatal(err)
	}
	var r restRepository
	if err := json.Unmarshal(b, &r); err != nil {
		log.Fatal(err)
	}
	fmt.Printf("%#+v\n", r)
	fmt.Println(r.Archived)
}
```

</details>

**Comment by keegancsmith** [comment]
Seeing the output of the rest repository listing API may also help us. We don't list specific repositories, but rather list via the repositories endpoint. The customer seems to be using `repositoryQuery: public`, so this should be somewhat equivalent to what we actually do:

```shell
curl \
  -H "Authorization: token token3" \
  -H "Accept: application/vnd.github.jean-grey-preview+json,application/vnd.github.mercy-preview+json" \
  -H "Accept: application/vnd.github.machine-man-preview+json" \
  'https://github.example.com/api/v3/repositories?per_page=100' | jq .
```

Added the extra headers since we include them: https://github.com/sourcegraph/sourcegraph/blob/0a19e43b9ea405f2ba4aa54ce7aaf6a39a121bf4/internal/extsvc/github/client.go#L257

**Comment by DaedalusG** [comment]
Hey has this issue been resolved? Yelp ended up using a bunch of excludes in their search config and it caused some weird issues that resulted in one of their indexes not showing up in search. https://github.com/sourcegraph/customer/issues/242

We ended up removing the excludes config relevant to the particular missing repo and that repo now returns, but the issue has morphed into `archived:only` searches returning nothing and then later returning the archived repos.

They'd like to clean up the excludes at this point but, and I'm wondering if that seems like a it'd be safe now?

**Comment by keegancsmith** [comment]
I don't remember working on this further. cc @ryanslade @asdine

**Comment by dbgrigsby** [comment]
@DaedalusG  This is still an issue on Sourcegraph 3.2.3. The `"repositoryQuery": [ "public"]` search type does not return the archived status of the repository, while the `orgs: [...]` external configuration does. This is still a somewhat regular issue for us at Yelp internally.

**Comment by DaedalusG** [comment]
@dbgrigsby Thanks for the info. I'll relay it to the private issue I've opened on this. Additionally, if you'd like to work with me on getting this sorted out feel free to start a thread in yelps support slack channel or open a support ticket and tag me @warren

**Comment by sfllaw** [comment]
We also have `repositoryQuery: ["affiliated", "public"]` defined, in addition to `orgs: [...]`. It looks like the public repositoryQuery clobbers the archive information provided by orgs?

Would a workaround be to remove the `repositoryQuery` selector?

https://docs.sourcegraph.com/admin/external_service/github#selecting-repositories-for-code-search

**Comment by sfllaw** [comment]
This limitation is documented in https://docs.sourcegraph.com/admin/external_service/github#repositoryquery-public-does-not-return-archived-status-of-a-repo:

> ### “repositoryQuery”: [“public”] does not return archived status of a repo
>
> The `repositoryQuery` option `"public"` is valuable in that it allows sourcegraph to sync all public repositories, however, it does not return whether or not a repo is archived. This can result in archived repos appearing in normal search. You can see an example of what is returned by the GitHub API for a query to “public” [here](https://docs.github.com/en/rest/reference/repos#list-public-repositories).
>
> If you would like to sync all public repositories while omitting archived repos, consider generating a GitHub token with access to only public repositories, then use `respositoryQuery` with option `affiliated` and an `exclude` argument with option `public` as seen in the example below:
> 
> ```
> {
>     "url": "https://github.example.com",
>     "gitURLType": "http",
>     "repositoryPathPattern": "devs/{nameWithOwner}",
>     "repositoryQuery": [
>         "affiliated"
>     ],
>     "token": "TOKEN_WITH_PUBLIC_ACCESS",
>     "exclude": [
>         {
>             "archived": true
>         }
>     ]
> }
> ```

**Comment by sfllaw** [comment]
> Would a workaround be to remove the `repositoryQuery` selector?

That totally worked around the problem.

I suppose the proper solution is to remember which repositories came from `repositoryQuery` and double-check their archived status.

**Comment by sashaostrikov** [comment]
I believe this issue has been resolved in v4.1.0 with this [commit](https://github.com/sourcegraph/sourcegraph/pull/41461).
